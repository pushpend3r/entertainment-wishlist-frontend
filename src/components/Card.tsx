import BCard from "react-bootstrap/Card";
import { Movie, TVShow } from "../types";

interface Props extends Partial<Movie & TVShow> {
  onClick: (id: Movie["id"] | TVShow["id"]) => void;
}

function Card({
  id,
  name,
  posterUrl,
  releaseDate,
  overview,
  startAirDate,
  lastAirDate,
  onClick,
  inProduction,
}: Props) {
  const releaseYear = releaseDate ? new Date(releaseDate).getFullYear() : "";
  const startYear = startAirDate ? new Date(startAirDate).getFullYear() : "";
  const lastYear = lastAirDate ? new Date(lastAirDate).getFullYear() : "";

  const title = `${name} (${releaseYear || startYear + "-" + `${!inProduction ? lastYear : "continue"}`})`;

  return (
    <BCard style={{ cursor: "pointer" }} onClick={() => onClick(id!)}>
      <BCard.Img variant="top" src={posterUrl} className="rounded" />
      <BCard.Body>
        <BCard.Title>{title}</BCard.Title>
        <BCard.Text
          style={{
            overflow: "hidden",
            textOverflow: "ellipsis",
            display: "-webkit-box",
            WebkitLineClamp: 3,
            WebkitBoxOrient: "vertical",
          }}
          title={overview}
        >
          {overview}
        </BCard.Text>
      </BCard.Body>
    </BCard>
  );
}

export default Card;
