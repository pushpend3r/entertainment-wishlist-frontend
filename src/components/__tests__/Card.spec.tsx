import { render, screen } from "@testing-library/react";
import Card from "../Card";

// eslint-disable-next-line
const emptyFunc = (_id: any) => {};

describe("<Card />", () => {
  test("render", () => {
    const image = "https://placehold.co/600x400";
    const title = "sometitle";
    const overview = "someoverview";

    const { container } = render(
      <Card onClick={emptyFunc} name={title} posterUrl={image} overview={overview} id={0} />
    );
    expect(container).toMatchSnapshot();
  });

  test("show release year in title", () => {
    const releaseDate = "2023-01-01";
    const releaseYear = new Date(releaseDate).getFullYear();
    render(<Card onClick={emptyFunc} releaseDate={releaseDate} id={0} />);
    const title = screen.getByRole("heading", { name: new RegExp(String(releaseYear)) });
    expect(title).toBeInTheDocument();
  });

  test("show start air year in title", () => {
    const startAirDate = "2023-01-01";
    const startAirYear = new Date(startAirDate).getFullYear();
    render(<Card onClick={emptyFunc} startAirDate={startAirDate} id={0} />);
    const title = screen.getByRole("heading", { name: new RegExp(String(startAirYear)) });
    expect(title).toBeInTheDocument();
  });

  test("show end air year in title", () => {
    const lastAirDate = "2023-01-01";
    const lastAirYear = new Date(lastAirDate).getFullYear();
    render(<Card onClick={emptyFunc} lastAirDate={lastAirDate} id={0} />);
    const title = screen.getByRole("heading", { name: new RegExp(String(lastAirYear)) });
    expect(title).toBeInTheDocument();
  });

  test("don't show start air year if release year is in title", () => {
    const startAirDate = "2023-01-01";
    const startAirYear = new Date(startAirDate).getFullYear();

    const releaseDate = "2024-01-01";
    const releaseYear = new Date(releaseDate).getFullYear();

    render(<Card onClick={emptyFunc} startAirDate={startAirDate} releaseDate={releaseDate} id={0} />);
    const title = screen.getByRole("heading", { name: new RegExp(String(releaseYear)) });

    expect(title).not.toHaveTextContent(new RegExp(String(startAirYear)));
  });
});
