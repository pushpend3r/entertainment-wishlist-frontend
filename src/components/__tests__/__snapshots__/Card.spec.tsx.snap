// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Card /> > render 1`] = `
<div>
  <div
    class="card"
    style="cursor: pointer;"
  >
    <img
      class="card-img-top rounded"
      src="https://placehold.co/600x400"
    />
    <div
      class="card-body"
    >
      <h2
        class="h5 card-title"
      >
        sometitle (-)
      </h2>
      <p
        class="card-text"
        style="overflow: hidden; text-overflow: ellipsis; display: -webkit-box; -webkit-line-clamp: 3;"
        title="someoverview"
      >
        someoverview
      </p>
    </div>
  </div>
</div>
`;
